Add support for sending serial output to STDERR

latchdevel opened the pull request:
Hi.

I am a grateful user of this software, but I needed to be able to send serial output to **STDERR** instead of **STDOUT**.

I have added support to be able to select the output both at compile time and at runtime. And by the way, I have implemented the `flush()` method which was not there.

I hope it helps you.
Regards.

latchdevel committed:
Add support for sending serial output to STDERR

bxparks commented:
1) Can you not redirect the STDOUT to STDERR at the shell level?
```
$ ./epoxy-binary 1>&2
```

2) Why do you need to call `fsync(2)`? StdioSerial uses `write(2)` which is an OS-level call, which bypasses the stdio.h buffers as far as I know. As long as the power plug isn't pulled, that should be safe. My understanding is that fsync() is needed to push the data to the physical device (the hard drive). But on modern system, the SSD or hard drive has its own cache/buffer, which can lie when fsync() is called, so the data may not actually be safe under power failure.

latchdevel commented:
Hi, thank you very much for your attention.

1. The need for StdioSerial to write to STDERR is because I need to independently write to STDOUT.

I have implemented a class I call IOClient to simulate a network client that uses both STDIN and STDOUT streams, allowing me to run sketches that use Ethernet or Wifi from the console.

```
$ echo -n "GET / HTTP/1.0\r\n\r\n" | ./epoxy-binary`
```

This way I can handle at shell level the STDOUT output as process output and mute or send to a file the output of Serial.print when it is on STDERR.

I have also implemented the dummy headers Ethernet.h, WiFi.h and ESP8266WiFi.h, so that I can compile any sketch. And with a simple python program I transform the `epoxy-binary` so that it interacts in real time through the network.

I have to do one last review of this before publishing it.

```c++
/*
   IOClient class: Interactive Client over standard streams.

   The IOClient class implements the Client interface, 
   using standard input and output streams STDIN and STDOUT.

   This allows for interactive operations suitable for server-side 
   applications. This server mode can be easily extended to support 
   TCP/IP communication with additional software.

*/

#ifndef IOClient_h_
#define IOClient_h_

#include "Client.h"
#include "IPAddress.h"

// Internal buffer size
#ifndef IOCLIENT_BUFFER_SIZE
#define IOCLIENT_BUFFER_SIZE 4096 // 4 Kbyte
#endif

class IOClient : public Client {
   public:
      // Basic TCP/IP connection simulation
      int connect(IPAddress ip, uint16_t port);
      int connect(const char *host, uint16_t port);
      uint8_t connected();
      operator bool();
      void stop();

      // Write to STDOUT
      size_t write(uint8_t value);
      size_t write(const uint8_t outputBuffer[], size_t outputBufferSize);
      void flush();

      // Get all available data from STDIN if the internal buffer has been completely read
      // Returns the number of bytes available to read from the internal buffer
      int available();

      // Read a byte from internal buffer
      int read();

      // Read a byte array from internal buffer
      int read(uint8_t outputBuffer[], size_t outputBufferSize);

      // Get the next byte of internal buffer without removing it from the buffer
      int peek();

   private:
      char readBuffer[IOCLIENT_BUFFER_SIZE] = {};  // Internal buffer to store STDIN data read
      uint16_t readBufferLength = 0;               // Internal buffer current length
      uint16_t readPosition = 0;                   // Internal buffer read position

      // Get how many bytes are available to read from STDIN
      int getBytesToRead();
};

#endif
```

2. Yes, you are right. At first I used `std::cout.flush();` but with write() it is not necessary.

bxparks commented:
**Question about IOClient**

Is your `IOClient` class intended to work *only* on a PC using EpoxyDuino? If it were intended to work on a real microcontroller, I would expect the `write()` functions to send the data over the network. For debugging, you'd probably want to pass in a `Serial` or `Serial2`  instance into `IOClient` using constructor injection, something like this:

```C++
class IOClient : public Stream {
public:
     IOClient(Print& logger) : 
       mLogger(logger),
       ....
      { ... }

private:
     Printer &mLogger;
};
```

My version of `IOClient` would work under both the real hardware and under EpoxyDuino. However I guess we still need the ability to create multiple instances of `StdioSerial` which are bound to different file descriptors.

**Proposed Alternative StdioSerial**

I'd like to propose a simpler version of `StdioSerial`, which directly uses the Unix file descriptor, without going through your `enum class StdioOutput` abstraction layer:

```C++
class StdioSerial: public Stream {
  public:
    StdioSerial(int fd = STDOUT_FILENO) : outputFd(fd) {}

    void setOutputDescriptor(int fd) { outputFd = fd; }
  ...
  private:
    int outputFd;
};
```

By default, the `StdioSerial` would use `STDOUT_FILENO`. Using the integer file description seems more flexible, because:

1)) It allows you to create an instance of StdioSerial bound to STDERR, like this:

```C++
StdioSerial SerialErr(STDERR_FILENO);
```

2)) It allows `StdioSerial` to be bound to an explicit file created by `open()` or `creat()`:

```C++
StdioSerial SerialToFile;
...
int fd = creat("/tmp/somefile", 0777 );
SerialToFile.setOutputFd(fd);
```

**Performance Implications**

Using the low-level `write()` functions becomes inefficient for large amount of data. My assumption was that `StdioSerial` was emulating the hardware `Serial` port, so it needed to handle only relatively small amounts of data. I suppose if speed becomes a problem, we could create a variant of `StdioSerial` which used the `FILE*` file stream objects, and their corresponding `fprintf()` functions.

bxparks commented:
> [snip]
> In summary, the **IOClient class** writes with `::write(STDOUT_FILENO, ...)` and reads with `::read(STDIN_FILENO, ...)`.
> 
> Everything works fine until the `Serial.print()` in the sketch also writes to STDOUT_FILENO, corrupting the output of the IOClient.

Ok, so it's basically a fake client that works only on EpoxyDuino for testing purposes.

> For this reason I needed to be able to change the output stream of Serial.print() to STDERR.
> So I modified the StdioSerial class, and I thought it would be a good idea to share it in case it is useful to someone else.

My proposed solution for `StdioSerial` is even simpler than yours, and perfectly backwards compatible. If you don't want to implement it, it's small enough that I can do it very quickly myself. Let me know if you plan to abandon this PR.

> The alternative you mention makes a lot of sense, especially in sketches with similar input/output streams such as Serial1, Serial2, SerialUSB, and even SoftwareSerial.

Yes, that would be required if you wanted `IOClient` to work under both EpoxyDuino and a real microcontroller. But if you need IOClient to work only on EpoxyDuino, my solution would work just as well, with a smaller code change.

bxparks commented:
I have implemented my solution as [12711fb](https://github.com/bxparks/EpoxyDuino/commit/12711fb29c0817e4764806087e83ef2392446c97), before I forget the context of this conversation.
Closing this as Obsolete.

latchdevel commented:
Hi, I'm glad you were able to make progress on this issue.

But this implementation is incomplete, since it doesn't allow setting the output to STDERR at compile time, and dangerous, since it allows any file descriptor number to be set to write.

I think it needs to be reviewed, I'll think about it.

Best regards.

bxparks commented:
1)) I'm not sure what you mean by "at compile time", but if you call `Serial.setOutputFileDescriptor(STDERR_FILENO)` in the global `setup()` before doing anything with `Serial`, then it's the equivalent to binding `Serial` to STDERR at compile time.

2)) Being able to pass any integer as the file descriptor is not dangerous. In fact, it is more flexible. For example, in [examples/StdioSerialMultiple](https://github.com/bxparks/EpoxyDuino/tree/develop/examples/StdioSerialMultiple), suppose we change `Serial1` to be bound to `3` (instead of STDOUT_FILENO which is 1), and `Serial2` to be bound to `4` (instead of STDERR_FILENO which is 2), like this:
```C++
StdioSerial Serial1(3);
StdioSerial Serial2(4);
...
void setup(void) {
...
  Serial1.println(F("Printing to 3"));
  Serial2.println(F("Printing to 4"));
}
```

Running this:
```
$ ./StdioSerialMultiple.out
```
will print nothing, because 3 and 4 are not bound to anything, but this should be fine, there is no danger that I am aware of.

**However**, you now have the ability to redirect descriptor 3 and 4 into separate files at the shell level, like this:

```
$ ./StdioSerialMultiple.out 3> stream3.txt 4> stream4.txt
```
After execution, you have 2 files: 
- `stream3.txt` contains the string "Printing to 3", and
- `stream4.txt` contains the string "Printing to 4"

latchdevel commented:
### 1)) 
I mean being able to change the output from `Serial.print()` to **STDERR** at compile time, without having to modify the source code.

That is, without making any modifications to the source code, being able to compile in such a way that STDERR is used.

```c++
// Set default output when compiling
#ifndef STDERR_OUTPUT
#define SERIAL_OUTPUT StdioOutput::STDOUT
#else 
#define SERIAL_OUTPUT StdioOutput::STDERR
#endif
```

In my proposal, this is possible like as:

```sh
$ c++ -DSTDERR_OUTPUT -std=gnu++11 -o file.o -c file.cpp
```

### 2))
It can be dangerous, if the descriptor number was previously assigned:

 ```c++
int fd = open("file.txt", O_RDWR); // it could return the value 3 or 4 for the file descriptor
...
StdioSerial Serial1(3);
StdioSerial Serial2(4);
...
void setup(void) {
...
   Serial1.println(F("Printing to 3")); // it write to "file.txt"
   Serial2.println(F("Printing to 4"));
}
```

I haven't thought about it much yet, and while I agree with being able to instantiate `Serial` objects indicating the flow to use, I think this should be of type `FILE`, or even better `Stream`, in order to support input operations like **`Serial1.read()`**.

Best regards.

bxparks commented:
> ### 1))
> ```shell
> $ c++ -DSTDERR_OUTPUT -std=gnu++11 -o file.o -c file.cpp
> ```

Good point. I think that's an easy fix using a layer of indirection, e.g. `SERIAL_DESCRIPTOR` which can be overridden at the command line. I'll prepare a fix soon.

> ### 2))
> It can be dangerous, if the descriptor number was previously assigned:

I'm going to assume that if someone is overridding the Unix file descriptors, they know what they are doing. I'm just providing access to the standard Unix file descriptor that people can figure out how to use.

> I think this should be of type `FILE`, or even better `Stream`, in order to support input operations like **`Serial1.read()`**.

I'm going to push back on that as "not in scope" at this point. My intention for EpoxyDuino is to provide a minimal implementation of the Arduino API, just enough to compile many Arduino programs, and to emulate the Serial port just enough for some minimal debugging. I don't want to maintain something more complex than absolutely necessary.

bxparks commented:
It will be something like:

```C++
#ifndef SERIAL_OUTPUT_FILENO
#define SERIAL_OUTPUT_FILENO STDOUT_FILENO
#endif
```

Which can be overridden by:
```
$ c++ -DSERIAL_OUTPUT_FILENO=2 -std=gnu++11 -o file.o -c file.cpp
```

I am assuming that anyone doing this sort of overriding knows that STDOUT==1 and STDERR==2 like the back of their hands. Because if they don't know that, they can't take advantage of bash(1) redirection using the `1>`, `2>`, and `3>` syntax.

bxparks commented:
`SERIAL_OUTPUT_FILENO` implemented in [abfe161](https://github.com/bxparks/EpoxyDuino/commit/abfe1613e028e984dd533babc52c0a581a4ada90)
